src/Orleans.Serialization/Utilities/FieldAccessor.cs

```diff
@@ -25,102 +33,60 @@ public static class FieldAccessor
-        public static Delegate GetReferenceSetter(FieldInfo field)
-        {
-            var delegateType = typeof(Action<,>).MakeGenericType(field.DeclaringType, field.FieldType);
-            return GetSetDelegate(field, delegateType, new[] { field.DeclaringType, field.FieldType });
-        }
+        public static Delegate GetReferenceSetter(FieldInfo field) => GetSetter(field, false);
 
         /// <summary>
         /// Returns a delegate to set the value of this field for an instance.
         /// </summary>
         /// <returns>A delegate to set the value of this field for an instance.</returns>
-        public static Delegate GetValueSetter(FieldInfo field)
-        {
-            var declaringType = field.DeclaringType;
-            if (declaringType is null)
-            {
-                throw new InvalidOperationException("Field " + field.Name + " does not have a declaring type.");
-            }
-
-            // Value types need to be passed by-ref.
-            var parameterTypes = new[] { declaringType.MakeByRefType(), field.FieldType };
-            var delegateType = typeof(ValueTypeSetter<,>).MakeGenericType(field.DeclaringType, field.FieldType);
-
-            return GetSetDelegate(field, delegateType, parameterTypes);
-        }
+        public static Delegate GetValueSetter(FieldInfo field) => GetSetter(field, true);
 
-        /// <summary>
-        /// Returns a delegate to set the value of a specified field.
-        /// </summary>
-        /// <param name="field">
-        /// The field.
-        /// </param>
-        /// <param name="delegateType">The delegate type.</param>
-        /// <param name="parameterTypes">The parameter types.</param>
-        /// <returns>A delegate to set the value of a specified field.</returns>
-        private static Delegate GetSetDelegate(FieldInfo field, Type delegateType, Type[] parameterTypes)
+        private static Delegate GetSetter(FieldInfo field, bool byref)
         {
-            var declaringType = field.DeclaringType;
-            if (declaringType is null)
-            {
-                throw new InvalidOperationException("Field " + field.Name + " does not have a declaring type.");
-            }
+            var declaringType = field.DeclaringType ?? throw new InvalidOperationException($"Field {field.Name} does not have a declaring type.");
+            var parameterTypes = new[] { typeof(object), byref ? declaringType.MakeByRefType() : declaringType, field.FieldType };
 
-            // Create a method to hold the generated IL.
             var method = new DynamicMethod(field.Name + "Set", null, parameterTypes, typeof(FieldAccessor).Module, true);
 
-            // Emit IL to return the value of the Transaction property.
             var emitter = method.GetILGenerator();
-            emitter.Emit(OpCodes.Ldarg_0);
+            // arg0 is unused for better delegate performance (avoids argument shuffling thunk)
```

Comment:
Could you explain this comment to me, why does this avoid argument shuffling? If you have links to other discussions/content on the matter, I'd be interested to read them

Reply:
It seems like a JIT deficiency

Reply:
There's a very detailed explanation here: https://mattwarren.org/2019/09/26/Stubs-in-the-.NET-Runtime/#delegates

But it's not a JIT limitation - the JIT can't know what kind of a delegate/target it's going to get when it generates the code for a delegate invocation callsite. So if we have a delegate `Action<A, B>` then the callsite will be `ld delegate; ld arg1; ld arg2; callvirt Action<A, B>.Invoke;` and the actual delegate target is a `static void Method(A a, B b);` then there's a parameter count mismatch and the delegate is created with an argument shuffling thunk that will basically do `#ignore arg0; ldarg.1; ldarg.2; call Method(A, B);`.

However, if the delegate target is an instance method (or equivalent), then there is no need for argument shuffling, just a single `mov ecx, [ecx + Delegate._target]` to replace the delegate instance value with the target instance value (null in our case).

Reply:
Great explanation, thank you

Reply:
If I understand correctly, your change will result in case 4 (Static closed) from this table:

```
// DELEGATE KINDS TABLE
//
//                                  _target         _methodPtr              _methodPtrAux       _invocationList     _invocationCount
//
// 1- Instance closed               'this' ptr      target method           null                null                0
// 2- Instance open non-virt        delegate        shuffle thunk           target method       null                0
// 3- Instance open virtual         delegate        Virtual-stub dispatch   method id           null                0
// 4- Static closed                 first arg       target method           null                null                0
// 5- Static closed (special sig)   delegate        specialSig thunk        target method       first arg           0
// 6- Static opened                 delegate        shuffle thunk           target method       null                0
// 7- Secure                        delegate        call thunk              MethodDesc (frame)  target delegate     creator assembly 
//
// Delegate invoke arg count == target method arg count - 2, 3, 6
// Delegate invoke arg count == 1 + target method arg count - 1, 4, 5
//
// 1, 4     - MulticastDelegate.ctor1 (simply assign _target and _methodPtr)
// 5        - MulticastDelegate.ctor2 (see table, takes 3 args)
// 2, 6     - MulticastDelegate.ctor3 (take shuffle thunk)
// 3        - MulticastDelegate.ctor4 (take shuffle thunk, retrieve MethodDesc) ???
//
// 7 - Needs special handling
```